engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/model/validation/StraightThroughProcessingLoopValidator.java

```diff
@@ -0,0 +1,222 @@
+/*
+ * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
+ * one or more contributor license agreements. See the NOTICE file distributed
+ * with this work for additional information regarding copyright ownership.
+ * Licensed under the Zeebe Community License 1.1. You may not use this file
+ * except in compliance with the Zeebe Community License 1.1.
+ */
+package io.camunda.zeebe.engine.processing.deployment.model.validation;
+
+import io.camunda.zeebe.engine.processing.common.Failure;
+import io.camunda.zeebe.engine.processing.deployment.model.element.AbstractFlowElement;
+import io.camunda.zeebe.engine.processing.deployment.model.element.ExecutableFlowNode;
+import io.camunda.zeebe.engine.processing.deployment.model.element.ExecutableProcess;
+import io.camunda.zeebe.engine.processing.deployment.model.element.ExecutableSequenceFlow;
+import io.camunda.zeebe.protocol.impl.record.value.deployment.DeploymentResource;
+import io.camunda.zeebe.protocol.record.value.BpmnElementType;
+import io.camunda.zeebe.util.Either;
+import io.camunda.zeebe.util.buffer.BufferUtil;
+import java.io.StringWriter;
+import java.util.ArrayList;
+import java.util.EnumSet;
+import java.util.LinkedList;
+import java.util.List;
+
+public final class StraightThroughProcessingLoopValidator {
+
+  private static final EnumSet<BpmnElementType> STRAIGHT_THROUGH_PROCESSING_ELEMENT_TYPES =
+      EnumSet.of(
+          BpmnElementType.MANUAL_TASK,
+          BpmnElementType.TASK,
+          BpmnElementType.EXCLUSIVE_GATEWAY,
+          BpmnElementType.INCLUSIVE_GATEWAY,
+          BpmnElementType.PARALLEL_GATEWAY);
+  private static final EnumSet<BpmnElementType> WHITELISTED_ELEMENT_TYPES =
+      EnumSet.of(
+          BpmnElementType.EXCLUSIVE_GATEWAY,
+          BpmnElementType.INCLUSIVE_GATEWAY,
+          BpmnElementType.PARALLEL_GATEWAY);
+
+  /**
+   * Validates a list of processes for straight-through processing loops. These are loops of
+   * elements that don't have any execution logic, such as undefined tasks and manual tasks.
+   *
+   * @param resource the resource file that's getting validated
+   * @param executableProcesses the list of processes in this resource
+   * @return an Either of which the left contains the failure message if any loops have been found
+   */
+  public static Either<Failure, ?> validate(
```

Comment:
:wrench: I guess an optional would be enough then here?

Reply:
Using an `Optional` is. bit awkward. It is definitely possible, but all the other validations work with `Eithers` as well. If I return an `Optional` here it means I'm gonna have to map it to an `Either` again where this gets called.

Reply:
Ah ok so the API expect either, wasn't aware. Felt just a bit weird since you will not return anything else.